BCP00301Test.py

```diff
@@ -117,3 +120,140 @@ def test_02(self, test):
+
+        ret = self.perform_test_ssl(test, ["-S"])
+        # Known issue: If the OCSP URL is unreachable testssl generates invalid JSON and exits with a non-zero code
+        if ret != 0:
+            return test.FAIL("Unable to test. See the console for further information.")
+        else:
+            common_name = None
+            with open(TMPFILE) as tls_data:
+                tls_data = json.load(tls_data)
+                for report in tls_data:
+                    if report["id"].split()[0] == "cert_commonName":
+                        common_name = report["finding"]
+                        try:
+                            ipaddress.ip_address(report["finding"])
+                            return test.WARNING("CN is an IP address: {}".format(report["finding"]))
+                        except ValueError:
+                            pass
+                    elif report["id"].split()[0] == "cert_subjectAltName":
+                        if report["finding"].startswith("No SAN"):
+                            return test.WARNING("No SAN was found in the certificate")
+                        else:
+                            alt_names = report["finding"].split()
+                            if common_name not in alt_names:
+                                return test.FAIL("CN {} was not found in the SANs".format(common_name))
+                            for name in alt_names:
+                                try:
+                                    ipaddress.ip_address(name)
+                                    return test.WARNING("SAN is an IP address: {}".format(name))
+                                except ValueError:
+                                    pass
+
+            if common_name is None:
+                return test.UNCLEAR("Unable to find CN in the testssl report")
+
+            return test.PASS()
+
+    def test_04_hsts(self, test):
+        """HSTS Header"""
+
+        ret = self.perform_test_ssl(test, ["-h"])
+        if ret != 0:
+            return test.FAIL("Unable to test. See the console for further information.")
+        else:
+            hsts_supported = False
+            with open(TMPFILE) as tls_data:
+                tls_data = json.load(tls_data)
+                for report in tls_data:
+                    if report["id"] == "HSTS":
+                        if report["severity"] == "OK":
+                            hsts_supported = True
+                        elif report["finding"] != "not offered":
+                            hsts_supported = report["finding"]
+            if hsts_supported is True:
+                return test.PASS()
+            elif hsts_supported is False:
+                return test.WARNING("Strict Transport Security (HSTS) should be supported")
+            else:
+                return test.FAIL("Error in HSTS header: {}".format(hsts_supported))
+
+    def test_05_revocation(self, test):
+        """Certificate revocation method is available"""

```

Comment:
Does this test do anything with regards to revocation lists? The spec says there SHALL be a way to revoke certs and the server SHOULD support OCSP, so a server with a revocation list should get a warning I guess.

Reply:
Yeah, the idea was for this one to catch the SHALL case so ensure there was some form of revocation, but not specifically CRLs. The next test then checks explicitly for OCSP. I could add another test for CRLs explicitly, but I hadn't for now as I don't believe the spec requires it.